File: packages/ai-service/app/vision.py
"""Photograph a thing, get a cue back.

    from app import vision
    vision.analyze(tenant="gap", kind="sku", image_base64=..., mime="image/jpeg",
                   note=None, crm=crm)   # -> {"lines": [...], "meta": [...]}

The G2 has no camera and that is a standing decision. The *phone* running the
plugin does, and the Even SDK hands it to us (`captureImageFromCamera` /
`pickImageFromAlbum`), so an associate can point it at a SKU tag or a broken
part and read the answer in the glass.

**Objects. Never people.**
--------------------------
This capture path is for a swing tag, a barcode, a serial plate, a snapped
bracket. It is not for a face, and there is nothing here that could become
one: no face detection, no person matching, no "who is this customer" — not
now and not as a follow-up. The provider prompt refuses a photograph whose
subject is a person and this module turns that refusal into a cue that says so.
The boundary is written here, in the Console copy and in the plugin's camera
permission description, because the standing camera-free, opt-in-only,
never-passive-matching decision is what gets this product into stores, and a
rule that lives in one person's memory is a rule that quietly moves.

**The image is never stored.**
------------------------------
Not to disk, not to the database, not in a log line, not in an analytics row.
It arrives as base64 in a request body, is passed once to the vision provider,
and goes out of scope with the request. Nothing in this module opens a file or
takes a database connection, and nothing prints the payload. This is the most
sensitive thing the service touches and the defence is that there is nothing
to find: an image that is never written cannot leak.

**A SKU answer comes from records, not from the model.**
--------------------------------------------------------
The model reads characters off a tag. That is all it is trusted to do. The
code it read is looked up in the floor inventory we actually hold, and the
price, location and count on the glass come from that record. If the code
matches nothing, the answer is that it matches nothing — which is a real
answer an associate can act on, and is the only honest one. The standing rule
is that answers are grounded in records we hold and never a model's guess
about stock; a vision path that prints a price the model thought it saw breaks
the most important promise in the product.

And "we hold no inventory records" is kept distinct from "we don't carry it".
`crm.floor_inventory()` is called bare elsewhere in this service, so an empty
feed and a floor that carries nothing are indistinguishable there — a live
store whose product query failed would tell an associate, confidently, that a
jean in their hand is not carried. This module refuses to make that mistake:
no records means we say we cannot confirm.
"""
from __future__ import annotations

import base64
import binascii
import re

from . import cue as cue_format
from .llm import VisionUnsupported, get_provider

#: What a capture can be about. `sku` resolves against inventory; `part` is a
#: field-service identification, which nothing on the floor can confirm.
KINDS = ("sku", "part")

#: Formats a phone actually produces and a vision model actually reads. HEIC is
#: absent on purpose: iOS converts on export through the SDK, and accepting a
#: container we cannot decode means a failure the associate can do nothing about.
ALLOWED_MIME = ("image/jpeg", "image/png", "image/webp")

#: 3 MB of image, decoded.
#:
#: A swing tag or a serial plate is legible at 1600px on the long edge, which
#: is a 200–600 KB JPEG; 3 MB leaves room for a 4000px capture of a part in bad
#: light. What it does not leave room for is the 12 MP original a phone hands
#: you if nobody downscales it, and that is the point — the client is expected
#: to resize before upload, and a limit that quietly accepts the unresized
#: version is a limit that trains nobody. Uploading four megabytes over shop
#: wifi also costs the associate the seconds this feature exists to save.
MAX_IMAGE_BYTES = 3 * 1024 * 1024

#: Base64 inflates by 4/3. Checked against the string *before* decoding, so a
#: caller who posts a 40 MB payload is refused without materialising it.
MAX_BASE64_CHARS = (MAX_IMAGE_BYTES + 2) // 3 * 4


class VisionRefused(Exception):
    """A capture this endpoint will not process, and why.

    Carries the HTTP status the route should return. Deliberately not an
    `HTTPException`: everything below is testable without a web framework, and
    the translation happens in exactly one place in `main.py`.
    """

    def __init__(self, status: int, detail: str):
        super().__init__(detail)
        self.status = status
        self.detail = detail


#: Asked of the model, per kind. Short, because the system prompt in `llm.py`
#: carries the rules (objects only, no invented facts, NONE when unreadable).
QUESTIONS = {
    "sku": ("Read the product identifier on this tag or label — the SKU, style "
            "number or barcode digits. Reply with that code alone."),
    "part": ("Name the object or part in this photograph in one short phrase, "
             "plus any model or serial number printed on it."),
}

#: The model's two reserved replies.
NO_READING = "NONE"
NOT_AN_OBJECT = "NO_SUBJECT"


def analyze(*, tenant: str, kind: str, image_base64: str, mime: str,
            note: str | None, crm) -> dict:
    """One photograph in, one cue out.

    Returns exactly what `cue.build` produces — `{"lines": [...], "meta": [...]}`
    — plus a few scalars the lens ignores, so this renders through the existing
    lens path with no new card type.
    """
    kind = (kind or "").strip().lower()
    if kind not in KINDS:
        raise VisionRefused(400, f"kind must be one of: {', '.join(KINDS)}")

    mime = (mime or "").strip().lower()
    if mime not in ALLOWED_MIME:
        raise VisionRefused(
            415, f"'{mime or 'unset'}' is not a supported image type. Send one "
                 f"of: {', '.join(ALLOWED_MIME)}.")

    payload = _checked_payload(image_base64)
    provider = get_provider()

    question = QUESTIONS[kind]
    if note:
        # Spoken context, clipped. It reaches the model as part of the question
        # and is not retained here any more than the image is.
        question = f"{question} The person taking it said: {str(note)[:200]}"

    try:
        reading = (provider.read_image(payload, mime, question) or "").strip()
    except VisionUnsupported as e:
        # An operator problem, not an associate problem: name the env var.
        raise VisionRefused(503, str(e))
    except Exception as e:
        # A model outage must not read as broken hardware. The associate
        # pressed a button and is looking at the glass, so the glass says what
        # happened — same posture as the voice path.
        return _out(kind, tenant, provider.name, grounded=False, sku=None,
                    lines=["COULD NOT READ IT", "TRY AGAIN IN A MOMENT"],
                    meta=[_short_reason(e)])

    if reading.upper().startswith(NOT_AN_OBJECT):
        # The boundary, as behaviour rather than as a comment.
        return _out(kind, tenant, provider.name, grounded=False, sku=None,
                    lines=["PHOTOGRAPH THE ITEM", "NOT PEOPLE"],
                    meta=["OBJECTS"])

    if kind == "sku":
        return _sku_answer(tenant, provider.name, reading, crm)
    return _part_answer(tenant, provider.name, reading, crm)


# --- the payload -------------------------------------------------------------

def _checked_payload(image_base64: str) -> str:
    """The base64 string, or a refusal that says what the ceiling is."""
    payload = (image_base64 or "").strip()
    if not payload:
        raise VisionRefused(400, "image_base64 is required")

    if len(payload) > MAX_BASE64_CHARS:
        raise VisionRefused(413, _too_big(len(payload) * 3 // 4))

    try:
        raw = base64.b64decode(payload, validate=True)
    except (binascii.Error, ValueError):
        raise VisionRefused(400, "image_base64 is not valid base64")

    size = len(raw)
    # Measured, then dropped. The decoded image has no reason to outlive this
    # function — the provider is handed the base64 the caller already sent.
    del raw
    if size > MAX_IMAGE_BYTES:
        raise VisionRefused(413, _too_big(size))
    if size < 1024:
        raise VisionRefused(400, "That image is too small to read anything from.")
    return payload


def _too_big(size: int) -> str:
    return (f"Image is {size / 1_048_576:.1f} MB; the limit is "
            f"{MAX_IMAGE_BYTES / 1_048_576:.0f} MB. Capture or resize to about "
            f"1600px on the long edge — a SKU tag reads fine at that size and "
            f"uploads in a fraction of the time.")


def _short_reason(e: Exception) -> str:
    return cue_format.glassify(type(e).__name__, cue_format.FACT_CHARS)


# --- SKU: grounded in what the floor actually holds ---------------------------

def _sku_answer(tenant: str, provider_name: str, reading: str, crm) -> dict:
    code = _identifier(reading)
    items, have_records = _floor(crm)

    if not have_records:
        # Empty and "not carried" are not the same claim, and this is the one
        # place the difference is load-bearing: an associate holding a jean
        # must not be told the store does not sell it because a product query
        # came back empty.
        return _out("sku", tenant, provider_name, grounded=False, sku=code,
                    lines=["NO FLOOR RECORDS LOADED",
                           f"CANT CONFIRM {code}" if code else "CANT CONFIRM IT",
                           "CHECK THE POS"],
                    meta=[])

    if not code:
        return _out("sku", tenant, provider_name, grounded=False, sku=None,
                    lines=["NO CODE READ", "MOVE CLOSER AND RETRY"], meta=[])

    item = _match(code, items)
    if item is None:
        # A real answer. The code was read and this floor has no record of it —
        # wrong store, an old tag, a transfer that never landed. Saying so beats
        # inventing a product, and beats a shrug.
        return _out("sku", tenant, provider_name, grounded=True, sku=code,
                    lines=["NOT IN FLOOR RECORDS", f"CODE {code}", "CHECK THE POS"],
                    meta=[])

    stock = item.get("stock")
    lines = [
        str(item.get("name") or code),
        f"{cue_format.money(item.get('price'))} · {item.get('location') or 'FLOOR'}",
    ]
    if isinstance(stock, int):
        lines.append(f"{stock} ON HAND")
    # Nothing in the meta strip: every fact worth having is already one of the
    # three lines, and cue.py's rule is that an empty strip is honest while a
    # padded one trains people to stop reading it.
    return _out("sku", tenant, provider_name, grounded=True,
                sku=item.get("sku") or code, lines=lines, meta=[])


def _identifier(reading: str) -> str | None:
    """The code the model read, or None if it read nothing usable.

    Kept strict on purpose: a model that returns a sentence has not read a code,
    and treating the sentence as one produces a lookup miss that reads to the
    associate as "we don't stock it".
    """
    text = (reading or "").strip().strip(".").strip()
    if not text or text.upper().startswith(NO_READING):
        return None
    # Models like to prefix an answer ("SKU: GAP-DNM-0498"), so take the last
    # token that looks like a code rather than the whole reply. A code contains
    # a digit — without that rule a model that answered in a sentence hands
    # "are" to the lookup, which misses, and a miss is printed to the associate
    # as "not in floor records". A sentence is not a failed lookup; it is a
    # failed read, and those want different words on the glass.
    candidates = [t for t in re.split(r"[\s,;:]+", text)
                  if re.fullmatch(r"[A-Za-z0-9][A-Za-z0-9\-_/.]{2,31}", t)
                  and any(c.isdigit() for c in t)]
    if not candidates:
        return None
    return candidates[-1].upper()


def _floor(crm) -> tuple[list[dict], bool]:
    """(items, we-actually-have-records).

    The second value is the whole reason this helper exists. A bare
    `crm.floor_inventory()` returns `[]` for a store with nothing on the floor,
    for a provider that failed, and for a tenant whose feed has not synced —
    three different situations that must not produce the same sentence.
    """
    try:
        items = crm.floor_inventory() or []
    except Exception:
        return [], False
    return list(items), bool(items)


def _norm_code(value) -> str:
    return re.sub(r"[^A-Z0-9]", "", str(value or "").upper())


def _match(code: str, items: list[dict]) -> dict | None:
    """Find the floor record this code refers to, or nothing.

    Exact first, then containment in either direction — a barcode carries the
    style number inside a longer string, and a shelf label often prints a
    prefix of it. Containment is only allowed above six characters, because
    below that it stops being an identifier and starts being a coincidence.
    """
    want = _norm_code(code)
    if not want:
        return None
    normalised = [(i, _norm_code(i.get("sku"))) for i in items]
    for item, sku in normalised:
        if sku and sku == want:
            return item
    if len(want) >= 6:
        for item, sku in normalised:
            if sku and len(sku) >= 6 and (sku in want or want in sku):
                return item
    return None


# --- part: a reading, labelled as a reading ----------------------------------

def _part_answer(tenant: str, provider_name: str, reading: str, crm) -> dict:
    """What the model saw, marked as what the model saw.

    Nothing on a retail floor can confirm a cracked hinge, so this branch has
    no record to be grounded in — and it says so in the meta strip rather than
    printing a model's sentence with the same authority as a stock count. If
    the reading does happen to name something the floor carries, the record's
    own facts go on the second line, and those are grounded.
    """
    if not reading or reading.upper().startswith(NO_READING):
        return _out("part", tenant, provider_name, grounded=False, sku=None,
                    lines=["COULDNT IDENTIFY IT", "TRY A CLOSER PHOTO"], meta=[])

    items, have_records = _floor(crm)
    item = _match(_identifier(reading) or "", items) if have_records else None

    lines = [reading]
    if item is not None:
        lines.append(f"{cue_format.money(item.get('price'))} · "
                     f"{item.get('location') or 'FLOOR'}")
    return _out("part", tenant, provider_name, grounded=False,
                sku=(item or {}).get("sku"), lines=lines,
                # The one fact worth spending the strip on: this sentence came
                # from a model looking at a picture, not from a record.
                meta=["MODEL READ"])


# --- shape -------------------------------------------------------------------

def _out(kind: str, tenant: str, provider_name: str, *, grounded: bool,
         sku: str | None, lines: list[str], meta: list[str]) -> dict:
    """`cue.build`'s shape, plus scalars the lens ignores.

    Full-frame width: a vision answer has no fact rail beside it, so it gets
    the 33-character budget rather than the guest card's 21.
    """
    cue = cue_format.build(lines, meta)
    return {
        **cue,
        "kind": kind,
        "tenant": (tenant or "gap").lower(),
        # True when every fact on the glass came out of a record we hold.
        "grounded": grounded,
        "sku": sku,
        "provider": provider_name,
    }

File: packages/glasses-plugin/src/vision.ts
/**
 * CueSea vision — photograph a thing with the phone, read the answer in-lens.
 *
 * ## The boundary, in plain words
 *
 * **The camera photographs things. It never photographs people.**
 *
 * It is for a SKU tag an associate cannot read, a washed-out care label, a
 * broken buckle they need the part number for. It is not for faces. There is
 * no face detection here, no person matching, no identification of anybody
 * from an image, and none of that is a setting somebody can turn on: the
 * plugin sends one photo to CueSea's own service and asks it what the object
 * is. If a person is in the frame by accident, nothing in this system looks
 * for them. That sentence is repeated in the `camera` permission description
 * in `app.json`, where the associate and the store can read it, and it should
 * be repeated anywhere this feature grows.
 *
 * The G2 has no camera and that stays true. This is the phone in the
 * associate's pocket. The glasses' only role is to show that something is
 * happening and then to show the answer.
 *
 * ## Which surface starts it
 *
 * The phone page, not the glasses. Aiming a camera is the most hands-and-eyes
 * task in the product — you are holding the phone and looking at it to frame a
 * swing tag whatever we do — so a glasses affordance only adds a step to
 * something that cannot be finished without the phone. The glasses' grammar is
 * no buttons and no chrome; their four input events are all spoken for.
 *
 * Eyes down to shoot, eyes up to read the answer while you are still talking
 * to the customer. The readout stays in the lens: a phone showing a result the
 * glasses do not is the failure this product exists to avoid.
 *
 * ## Why it looks the way it does
 *
 * A photo and a network round trip is slower than voice, and voice at ~3s is
 * already at the edge of what someone will hold still for. So the glass shows
 * its stage and the seconds elapsed, in exactly the shape `voice.ts` uses for
 * listening — one pattern for "the machine is working", not two.
 *
 * And every failure says *which* failure. "Something went wrong" sends an
 * associate to a manager for a thing they could have fixed themselves, and
 * only one of the failures below is actually a manager's problem.
 *
 * Nothing here stores the image, and nothing logs the base64. The bytes exist
 * for one POST and are dropped.
 */
import type { CaptureResult, GlassesBridge } from "./bridge";

/**
 * What this store has turned on.
 *
 * Fetched from the realtime server at boot — the plugin is a static bundle and
 * cannot hold a service key, so every capability question is proxied like
 * everything else.
 */
export interface Capabilities {
  /** Whether the server found a tenant with this slug at all. Not a switch —
   *  the switches already fail closed. This is so the page can say *which*
   *  nothing it is looking at, instead of rendering a typo as a real store
   *  with the camera switched off. */
  known: boolean;
  camera_capture: boolean;
  voice: boolean;
  floor_comms: boolean;
}

/**
 * Everything off.
 *
 * The default is closed, and the camera is why: an affordance that appears
 * because a fetch timed out is a camera control on the face of an associate
 * whose store declined the camera. Failing the other way costs a feature for
 * as long as the network is down, which is recoverable and visible.
 */
export const NO_CAPABILITIES: Capabilities = { known: false,
  camera_capture: false,
  voice: false,
  floor_comms: false,
};

/** Don't hold the HUD hostage to a slow store network. */
const CAPABILITIES_TIMEOUT_MS = 2_500;

/**
 * Ask the server what this tenant may do. **Never rejects, never throws** —
 * every failure resolves to `NO_CAPABILITIES`, because a caller that has to
 * remember to catch is a caller that will one day forget and fail open.
 *
 * A malformed or partial body is a failure too: a flag that isn't a boolean
 * has not said yes, and `Boolean(undefined)` quietly reading as "off" is only
 * safe by accident. It is asserted rather than relied on.
 */
export async function fetchCapabilities(
  baseUrl: string,
  tenant: string,
  fetchImpl: typeof fetch = fetch,
): Promise<Capabilities> {
  const timeout = new Promise<null>((resolve) =>
    setTimeout(() => resolve(null), CAPABILITIES_TIMEOUT_MS),
  );
  try {
    const res = await Promise.race([
      fetchImpl(`${baseUrl}/api/tenant/capabilities?tenant=${encodeURIComponent(tenant)}`),
      timeout,
    ]);
    if (!res || !res.ok) return { ...NO_CAPABILITIES };
    const body = await res.json();
    if (!body || typeof body !== "object") return { ...NO_CAPABILITIES };
    return {
      // An older service that predates this field answers `undefined`, which
      // reads as unknown — wrong, and it would paint every existing tenant as
      // a typo. Absent means "this service cannot tell me", and a service that
      // cannot tell us is not evidence the tenant is missing.
      known: body.known === undefined ? true : body.known === true,
      camera_capture: body.camera_capture === true,
      voice: body.voice === true,
      floor_comms: body.floor_comms === true,
    };
  } catch {
    return { ...NO_CAPABILITIES };
  }
}

/** What the service is being asked to recognise. */
export type VisionKind = "sku" | "part";

export interface CaptureControl {
  label: string;
  kind: VisionKind;
}

/**
 * The camera affordance on the phone page — **or nothing at all.**
 *
 * When the store has not turned the camera on these controls do not exist. Not
 * greyed out, not erroring when pressed, not present-but-refused: an associate
 * should never see a control their store declined, because a control that
 * exists and does not work is a support call, and a control that exists and
 * *shouldn't* is a conversation with a customer about why the phone is taking
 * pictures on the shop floor.
 *
 * The gate is a list rather than a flag on purpose. The page builds its card
 * from whatever this returns, so an empty array is an absent card by
 * construction — there is no branch where a control gets built and then hidden.
 *
 * Two controls rather than one because the service wants to know whether it is
 * reading a tag or identifying a part, and asking on the phone — where the
 * associate is already looking — is free. The alternative is guessing.
 *
 * Sentence case, in the page's sans face: this is a control a person presses,
 * not a line the machine says in the glass.
 */
export function captureControls(caps: Capabilities): CaptureControl[] {
  if (!caps?.camera_capture) return [];
  return [
    { label: "Scan a tag", kind: "sku" },
    { label: "Scan a part", kind: "part" },
  ];
}

/**
 * The cap on what we will put on the wire.
 *
 * A modern phone camera will hand back 3-8 MB, and base64 adds a third on top
 * of that; a store's wifi will sit on it for half a minute and the associate
 * is standing in front of somebody. Refusing early with a sentence they can
 * act on beats a timeout they cannot read.
 *
 * 3 MiB is the service's own ceiling on the decoded image, so this is the same
 * number rather than a looser one: a photo that is going to be refused should
 * be refused here, in the moment, instead of after an upload over shop wifi
 * that was never going to be accepted. A 413 is still handled — the two limits
 * living in two repositories is exactly how they drift apart — but it should
 * not be the way an associate finds out.
 */
export const MAX_IMAGE_BYTES = 3 * 1024 * 1024;

export type VisionState = "idle" | "capturing" | "analysing" | "answering";

/**
 * Every way this can end badly, each with its own sentence.
 *
 * The list is the point. "Something went wrong" would cover all seven, and an
 * associate reading it has no idea whether to retake the photo, walk to a
 * manager, or wait for the wifi.
 */
export type VisionFailure =
  | "cancelled"      // they backed out of the camera
  | "unsupported"    // this Even App has no camera API
  | "capture-failed" // the host tried and couldn't
  | "forbidden"      // 403: the tenant's camera is off
  | "too-large"      // the photo is bigger than we will send
  | "unrecognised"   // 200, but the service saw nothing it knows
  | "unreachable"    // the request never landed
  | "service-error"; // it landed and came back wrong

/**
 * Written to 21 characters, which is what a line holds beside the fact rail —
 * the tightest place any of these can land, and where they will land, because
 * an associate scanning a tag is usually mid-engagement with the rail up.
 */
export function failureCue(reason: VisionFailure, detail = ""): string[] {
  switch (reason) {
    case "cancelled":
      return ["NO PHOTO TAKEN", "SCAN AGAIN ON PHONE"];
    case "unsupported":
      return ["CAMERA UNAVAILABLE", "UPDATE THE EVEN APP"];
    case "capture-failed":
      return ["CAMERA DIDNT OPEN", "TRY THE PHONE CAMERA"];
    case "forbidden":
      // The one failure that really is a manager's problem, said plainly so
      // the walk is worth it.
      return ["CAMERA OFF IN STORE", "A MANAGER CAN ENABLE"];
    case "too-large":
      return ["PHOTO TOO LARGE", detail, "TRY A SMALLER PHOTO"].filter(Boolean);
    case "unrecognised":
      return ["NOTHING RECOGNISED", "TRY A CLOSER PHOTO"];
    case "unreachable":
      return ["CANT REACH CUESEA", "CHECK THE CONNECTION"];
    case "service-error":
      return ["LOOKUP FAILED", detail, "TRY AGAIN"].filter(Boolean);
  }
}

/** What `/api/vision/analyze` answers with — the shape `renderCue` takes. */
export interface VisionResult {
  lines?: string[];
  meta?: string[];
}

export interface VisionDeps {
  bridge: GlassesBridge;
  /** Where the realtime server lives. Vision is proxied through it like
   *  everything else; the plugin cannot hold the service key. */
  serverUrl: string;
  tenant: string;
  /** Same signature `voice.ts` uses — main.ts owns the page lifecycle, and a
   *  vision answer is a cue like any other. */
  render(lines: string[], meta: string[]): Promise<void>;
  log(message: string): void;
  onState?(state: VisionState): void;
  /** The answer's dwell expired, or it was dismissed. */
  onDone?(): void;
  fetchImpl?: typeof fetch;
}

/** How long an answer stays up. Matches voice's dwell — it is the same kind of
 *  answer and there is no reason for two numbers. */
export const ANSWER_DWELL_MS = 12_000;
/**
 * How often the working screen repaints its elapsed count.
 *
 * `voice.ts` repaints at 400ms because it is animating a level meter off live
 * audio. There is nothing live to show inside a fetch, so this counts whole
 * seconds and repaints twice a second — enough that the number is never more
 * than half a second stale, and half the traffic to the host.
 */
const TICK_MS = 500;

export class VisionController {
  private state: VisionState = "idle";
  private startedAt = 0;
  private tickTimer: ReturnType<typeof setInterval> | null = null;
  private dwellTimer: ReturnType<typeof setTimeout> | null = null;
  private stageLines: string[] = [];

  constructor(private deps: VisionDeps) {}

  get current(): VisionState {
    return this.state;
  }

  private setState(next: VisionState) {
    this.state = next;
    this.deps.onState?.(next);
  }

  /**
   * Capture → show that something is happening → POST → render the answer.
   *
   * Re-entrant by refusal: a second scan while one is in flight is ignored
   * rather than queued. Two photos racing to the same three lines is worse
   * than a press that did nothing.
   */
  async scan(kind: VisionKind): Promise<void> {
    if (this.state !== "idle" && this.state !== "answering") return;
    this.clearDwell();
    this.startedAt = Date.now();

    const thing = kind === "part" ? "PART" : "TAG";
    // Said before the camera opens, because the phone is about to take the
    // associate's attention and the glass is what they look back at.
    this.setState("capturing");
    this.stageLines = ["CAMERA OPEN ON PHONE", `PHOTOGRAPH THE ${thing}`];
    await this.paint();
    this.startTicking();

    const shot = await this.deps.bridge.captureImage();
    if (!shot.ok) {
      // Cancelling is not a failure worth four seconds of screen: it was a
      // deliberate act and the associate already knows what they did. It still
      // gets its own sentence, briefly, so the glass never just snaps back.
      if (shot.reason === "failed") this.deps.log(`vision: capture failed — ${shot.detail}`);
      return this.fail(
        shot.reason === "cancelled" ? "cancelled"
          : shot.reason === "unsupported" ? "unsupported"
            : "capture-failed",
        shot.reason === "cancelled" ? 2_500 : 5_000,
      );
    }

    const { asset } = shot;
    // `size` is the host's byte count for the encoded image. Guard on the
    // base64 length as well: it is what actually goes on the wire, and a host
    // that reports size as 0 (some do) would otherwise walk straight past this.
    const bytes = Math.max(asset.size || 0, Math.floor((asset.base64.length * 3) / 4));
    if (bytes > MAX_IMAGE_BYTES) {
      // Whole megabytes: the glass charset is [A-Z0-9 ·%$£€/+-] and has no
      // decimal point, so "5.2 MB" would arrive as "5 2 MB" — a number that
      // reads as two numbers.
      return this.fail("too-large", 5_000, `${Math.round(bytes / 1_000_000)} MB`);
    }
    // The size and the kind, never the bytes. There is no path from this
    // module to storage and none to the log.
    this.deps.log(`vision: ${kind} photo, ${Math.round(bytes / 1024)} KB — sending`);

    this.setState("analysing");
    this.stageLines = [kind === "part" ? "READING THE PART" : "READING THE TAG"];
    await this.paint();

    let res: Response | null = null;
    try {
      res = await (this.deps.fetchImpl ?? fetch)(`${this.deps.serverUrl}/api/vision/analyze`, {
        method: "POST",
        headers: { "content-type": "application/json" },
        body: JSON.stringify({
          tenant: this.deps.tenant,
          kind,
          image_base64: asset.base64,
          mime: asset.mimeType || "image/jpeg",
        }),
      });
    } catch (e) {
      // A thrown fetch is a request that never landed — no server, no wifi, no
      // DNS. Distinct from one that landed and answered badly, and distinct in
      // what the associate should do about it.
      this.deps.log(`vision: unreachable — ${String(e)}`);
      return this.fail("unreachable", 5_000);
    }

    if (res.status === 403) return this.fail("forbidden", 6_000);
    // The service's own size ceiling. The check above should have caught this
    // before the upload; if it did not, the two limits have drifted and the
    // associate still gets the sentence that tells them what to do about it,
    // rather than a number.
    if (res.status === 413) return this.fail("too-large", 5_000);
    if (!res.ok) return this.fail("service-error", 5_000, `ERROR ${res.status}`);

    let body: VisionResult | null = null;
    try {
      body = await res.json();
    } catch {
      body = null;
    }
    const lines = (body?.lines || []).filter(Boolean);
    if (!lines.length) return this.fail("unrecognised", 5_000);

    this.stopTicking();
    this.setState("answering");
    await this.deps.render(lines, body?.meta ?? []);
    this.deps.log(`vision: ${kind} answered in ${Date.now() - this.startedAt}ms`);
    this.scheduleDwell(ANSWER_DWELL_MS);
  }

  /** A press while a scan is up: clear it and go back to what was underneath.
   *  Returns whether it had anything to clear, so `main.ts` can fall through
   *  to the next meaning of a press when it didn't. */
  dismiss(): boolean {
    if (this.state === "idle") return false;
    // A capture in flight cannot be recalled — the phone owns the camera UI
    // and will answer eventually. Dropping our state means its answer lands on
    // an idle controller and is discarded, which is the right outcome.
    this.stopTicking();
    this.clearDwell();
    this.setState("idle");
    this.deps.onDone?.();
    return true;
  }

  private async fail(reason: VisionFailure, dwellMs: number, detail = "") {
    this.stopTicking();
    this.deps.log(`vision: ${reason}${detail ? ` (${detail})` : ""}`);
    this.setState("answering");
    await this.deps.render(failureCue(reason, detail), []);
    this.scheduleDwell(dwellMs);
  }

  /**
   * The working screen: the stage, and how long it has been going.
   *
   * The seconds are the honest half. A photo round trip is slower than voice
   * and there is no progress to report from inside a fetch — a counter at
   * least distinguishes "working" from "frozen on the last frame", which is
   * the thing the glass must never do.
   */
  private async paint() {
    // Whole seconds, floored. Not `toFixed(1)`: the glass charset is
    // [A-Z0-9 ·%$£€/+-] and has no full stop, so "3.4S" arrives as "3 4S" —
    // one number that reads as two. At the length of a photo round trip the
    // tenths were never worth reading anyway.
    const seconds = Math.floor((Date.now() - this.startedAt) / 1000);
    await this.deps.render(this.stageLines, [`${seconds}S`]);
  }

  private startTicking() {
    this.stopTicking();
    this.tickTimer = setInterval(() => void this.paint(), TICK_MS);
  }

  private stopTicking() {
    if (this.tickTimer) clearInterval(this.tickTimer);
    this.tickTimer = null;
  }

  private scheduleDwell(ms: number) {
    this.clearDwell();
    this.dwellTimer = setTimeout(() => {
      this.dwellTimer = null;
      this.setState("idle");
      this.deps.onDone?.();
    }, ms);
  }

  private clearDwell() {
    if (this.dwellTimer) clearTimeout(this.dwellTimer);
    this.dwellTimer = null;
  }
}
